src/main/scala/ws/WsConnection.scala

```diff
@@ -1,36 +1,72 @@
 package io.blindnet.dataaccess
 package ws
 
-import models.Namespace
+import models.Connector
 import redis.DataRequestRepository
 import ws.*
 
 import cats.data.*
 import cats.effect.*
 import cats.effect.std.Queue
+import fs2.{Pipe, Stream}
 import io.circe.*
 import io.circe.parser.*
 import io.circe.syntax.*
 
 import java.nio.ByteBuffer
 import java.util.UUID
+import scala.util.Try
+
+case class WsConnection(repos: Repositories, connector: Option[Connector], queue: Queue[IO, String]) {
+  def pipe(onTerminate: IO[Unit]): Pipe[IO, String, String] = (in: Stream[IO, String]) => {
+    // noneTerminate should theoretically be enough to detect disconnects, but the Stream will actually fail
+    // with an EOF error and therefore not emit a None.
+    // We are keeping the call to noneTerminate, in the unlikely event that it could actually not fail sometimes.
+    // unNoneTerminate prevents emitting None twice in this same unlikely event.
+    //
+    // See (*probably* related):
+    // https://github.com/http4s/blaze/issues/668
+    // https://github.com/softwaremill/adopt-tapir/pull/223
+    Stream.fromQueueUnterminated(queue, Int.MaxValue).merge(
+      in.noneTerminate
+        .handleErrorWith(_ => Stream(None))
+        .evalTap(_.map(receive).getOrElse(onTerminate))
+        .unNoneTerminate
+        .drain
+    )
+  }
+
+  private def receive(raw: String): IO[Unit] =
+    def parsePacket[T <: WsInPacket](): Option[(T, Option[(UUID, UUID)])] =
+      def parseUUID(j: Json): Option[UUID] = j.asString.flatMap(s => Try(UUID.fromString(s)).toOption)
 
-case class WsConnection(repos: Repositories, ns: Namespace, queue: Queue[IO, String]) {
-  def receive(raw: String): IO[Unit] =
-    def parsePacket[T <: WsInPacket](): Option[T] =
       for {
         payload <- parse(raw).toOption.flatMap(_.asObject)
         typ <- payload("typ").flatMap(_.asString)
         decoder <- WsInPacket.decoders.get(typ)
         packet <- payload("data").flatMap(_.as[T](decoder.asInstanceOf[Decoder[T]]).toOption)
-      } yield packet
+      } yield (packet,
+        payload("app_id").flatMap(parseUUID)
```

Comment:
Note: global connectors not deployed by us could lie about these data.

Reply:
With the fixed global conn token, I'm assuming DAC and global connectors are deployed by the same entity, so this should not be an issue.

I don't really see the point of self-hosting global connectors. You'd be better off deploying a custom one with the only one you need.